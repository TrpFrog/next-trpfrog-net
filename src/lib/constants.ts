export const LOREM_IPSUM = [
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit, ',
  'sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
  'Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
  'Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.',
  'Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
].join('')

export const SITE_NAME = 'つまみネット'

export const TWITTER_ID = 'TrpFrog'

export const TRPFROG_DIFFUSION_DEFAULT_UPDATE_HOURS = 3
export const TRPFROG_DIFFUSION_UPDATE_HOURS_EDGE_CONFIG_KEY =
  'trpfrog-diffusion-update-hours'

export const ATCODER_HIGHEST = 1596
export const WALKING_FARTHEST = '70.5km'

export const PRODUCTION_HOST = 'https://trpfrog.net'
export const DEVELOPMENT_HOST = 'http://localhost:3000'
export const CURRENT_HOST =
  process.env.NODE_ENV === 'production' ? PRODUCTION_HOST : DEVELOPMENT_HOST
export const HOST_URL =
  process.env.NODE_ENV === 'production'
    ? `https://${process.env.VERCEL_URL || 'trpfrog.net'}`
    : DEVELOPMENT_HOST

export const DEFAULT_BLOG_THUMBNAIL =
  'https://res.cloudinary.com/trpfrog/TwitterCard'
