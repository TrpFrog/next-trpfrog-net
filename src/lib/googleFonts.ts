// import {
//   BIZ_UDPGothic,
//   Comfortaa,
//   M_PLUS_Rounded_1c,
//   Noto_Sans_JP,
//   Noto_Sans_Mono,
//   Roboto
// } from "next/font/google";
// import {NextFontWithVariable} from "next/dist/compiled/@next/font";
//
// const mPlusRounded1c = M_PLUS_Rounded_1c({
//   subsets: ['latin'],
//   weight: ["100", "400", "700", "800"],
//   display: 'swap',
//   variable: '--font-m-plus-rounded-1c',
// })
//
// const comfortaa = Comfortaa({
//   subsets: ['latin'],
//   variable: '--font-comfortaa',
// })
//
// const roboto = Roboto({
//   subsets: ['latin'],
//   weight: ["400", "700"],
//   variable: '--font-roboto',
// })
//
// const notoSansJP = Noto_Sans_JP({
//   subsets: ['latin'],
//   weight: ["400", "700"],
//   variable: '--font-noto-sans-jp',
// })
//
// const notoSansMono = Noto_Sans_Mono({
//   subsets: ['latin'],
//   variable: '--font-noto-sans-mono',
// })
//
// const bizUdpGothic = BIZ_UDPGothic({
//   subsets: ['latin'],
//   weight: ["400", "700"],
//   variable: '--font-biz-udp-gothic',
// })
//
// const fonts: NextFontWithVariable[] = [
//   mPlusRounded1c,
//   comfortaa,
//   roboto,
//   notoSansJP,
//   notoSansMono,
//   bizUdpGothic,
// ]
//
// const fontVariables = fonts.map((font) => font.variable).join(' ');

// Temporary solution for the next/font issue
import styles from '@/components/GoogleFonts/index.module.scss'
const fontVariables = styles.temp_google_font_variables

export default fontVariables
