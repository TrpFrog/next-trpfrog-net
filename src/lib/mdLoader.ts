import fs from 'fs/promises'
import path from 'path'

import matter from 'gray-matter'
import { z } from 'zod'

export type MarkdownWithFrontmatter<T> = {
  filename: string
  metadata: T
  content: string
}

export async function readMarkdowns<T extends { date: Date }>(
  dirpath: string,
  schema: z.Schema<T>,
): Promise<MarkdownWithFrontmatter<T>[]> {
  const files = await fs.readdir(dirpath)
  const markdowns = files.filter(file => {
    return path.extname(file) === '.md'
  })
  const markdownsWithFrontmatter = await Promise.all(
    markdowns.map(async filename => {
      const file = await fs.readFile(path.join(dirpath, filename), 'utf8')
      const matterResult = matter(file)
      return {
        filename,
        metadata: schema.parse(matterResult.data),
        content: matterResult.content,
      } as const
    }),
  )
  return markdownsWithFrontmatter.sort((a, b) =>
    a.metadata.date < b.metadata.date ? 1 : -1,
  )
}
