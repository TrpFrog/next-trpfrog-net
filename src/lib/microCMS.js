import { createClient } from 'microcms-js-sdk'

import { env } from '@/env'

const client = createClient({
  serviceDomain: 'trpfrog',
  apiKey: env.MICRO_CMS_API_KEY,
})

export { client as microCMS }
