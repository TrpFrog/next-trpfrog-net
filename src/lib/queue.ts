export class Queue<T> {
  size: number
  items: T[]
  first: T | null
  last: T | null

  constructor() {
    this.size = 0
    this.items = []
    this.first = null
    this.last = null
  }

  push(item: T) {
    this.items.push(item)
    this.last = item
    this.first = this.items[0] ?? null
    this.size++
  }

  pop() {
    this.size = Math.max(0, this.size - 1)
    const removed = this.items.shift() ?? null
    this.last = this.items[this.size - 1] ?? null
    this.first = this.items[0] ?? null
    return removed
  }
}
