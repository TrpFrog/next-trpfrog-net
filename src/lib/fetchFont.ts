import { z } from 'zod'

import { createURL } from '@/lib/url'

import { env } from '@/env/server'

const FontInfoSchema = z.object({
  items: z
    .object({
      family: z.string(),
      files: z.record(z.string(), z.string().url()),
    })
    .array(),
})

export async function fetchFont(
  familyName: string,
  weight: string | number,
): Promise<ArrayBuffer> {
  if (!env.GOOGLE_FONTS_API_KEY) {
    throw new Error('GOOGLE_FONTS_API_KEY is not set')
  }

  if (typeof weight === 'number') {
    weight = weight.toString()
  }

  const endpoint = createURL(
    '/webfonts/v1/webfonts',
    'https://www.googleapis.com',
    {
      family: familyName,
      key: env.GOOGLE_FONTS_API_KEY,
    },
  )

  const rawGoogleFontsInfo = await fetch(endpoint).then(res => res.json())

  const googleFontsInfo = await FontInfoSchema.parseAsync(
    rawGoogleFontsInfo,
  ).catch(e => {
    console.error(`Endpoint: ${endpoint}`)
    console.error(`Response: ${JSON.stringify(rawGoogleFontsInfo, null, 2)}`)
    console.error(e)
    throw e
  })
  const url = googleFontsInfo.items[0].files[weight]

  return fetch(url, {
    cache: 'no-cache',
  }).then(res => res.arrayBuffer())
}
