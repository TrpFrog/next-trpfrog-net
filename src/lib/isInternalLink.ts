import { DEVELOPMENT_HOST, PRODUCTION_HOST } from '@/lib/constants'

export function isInternalLink(href: string): boolean {
  return ['/', '#', 'mailto:', PRODUCTION_HOST, DEVELOPMENT_HOST].some(prefix =>
    href.startsWith(prefix),
  )
}
