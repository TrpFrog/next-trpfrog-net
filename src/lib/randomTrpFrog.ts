import { HfInference } from '@huggingface/inference'
import { ChatOpenAI } from 'langchain/chat_models/openai'
import { AIMessage, HumanMessage, SystemMessage } from 'langchain/schema'
import { z } from 'zod'

import { getShuffledArray } from '@/lib/arrayHelpers'
const hf = new HfInference(process.env.HUGGINGFACE_TOKEN)

async function fetchRandomWords(amount: number) {
  const apiUrl = 'https://random-word-api.herokuapp.com'
  const response = await fetch(`${apiUrl}/word?number=${amount}&lang=en`, {
    next: {
      revalidate: 3, // seconds
    },
  })
  const words = z
    .string()
    .array()
    .length(amount)
    .parse(await response.json())
  return words
}

async function generateRandomTrpFrogPrompt(sourceWords: string[]) {
  // Few-shot learning
  const chatInput = [
    new SystemMessage(
      [
        'The user inputs some words.',
        'Now, there is a image-generation AI based on diffusion models called `trpfrog-diffusion`.',
        'One of your task is to generate a prompt for the AI to generate an image based on the words above.\n',
        "  - You don't have to use all the words, but you can't use any other words.\n",
        '  - You must not use sensitive (e.g. sexual, violent) words. This is the very important rule.\n',
        '  - The prompt should starts with "a photo of" and includes "trpfrog". ("trpfrog" is a subject of the photo) \n',
        '  - You must not generate a sensitive prompt. This is also the very important rule.\n',
        '\n',
        'Your task is process following tasks in order and answer it in JSON format:\n',
        '  "prompt": Generate a prompt for the AI to generate an image based on the words above in English.\n',
        '  "translated": Translate this prompt into plain Japanese. ',
        '("trpfrog" is "つまみさん" in Japanese,',
        ' and you should try to translate English-specific words into Japanese using カタカナ.)\n',
      ].join(''),
    ),
    new HumanMessage(
      getShuffledArray([...sourceWords.slice(1), 'running']).join(', '),
    ),
    new AIMessage(
      JSON.stringify({
        prompt: 'a photo of running trpfrog',
        translated: '走るつまみさんの画像',
      }),
    ),
    new HumanMessage(
      getShuffledArray([...sourceWords.slice(2), 'hat', 'sword']).join(', '),
    ),
    new AIMessage(
      JSON.stringify({
        prompt: 'a photo of trpfrog with a hat and a sword and a shield',
        translated: '帽子と剣と盾を持ったつまみさんの画像',
      }),
    ),
    new HumanMessage(sourceWords.join(', ')),
  ]

  const chat = new ChatOpenAI({ temperature: 0.3 })
  const reply = await chat.call(chatInput).then(res => res.content)
  return z
    .object({
      prompt: z.string().transform(s => s.trim().toLowerCase()),
      translated: z.string(),
    })
    .parse(JSON.parse(reply))
}

export async function generateTrpFrogImage(prompt: string) {
  const responseBlob = await hf.textToImage(
    {
      model: 'Prgckwb/trpfrog-diffusion',
      inputs: prompt,
    },
    {
      use_cache: false,
      wait_for_model: true,
      retry_on_error: true,
    },
  )
  const arrayBuffer = await responseBlob.arrayBuffer()
  const base64 = Buffer.from(arrayBuffer).toString('base64')

  // if sensitive image is generated, huggingface returns a black image.
  // Base64 of this image contains "ooooAKKKKACiiigA" pattern.
  const invalidImagePattern = /(ooooAKKKKACiiigA){10,}/

  return {
    base64,
    success: !invalidImagePattern.test(base64),
  }
}

type GeneratedPrompt = {
  prompt: string
  translated: string
}

export async function generateRandomPrompt(): Promise<GeneratedPrompt> {
  const words = await fetchRandomWords(10)
  return await generateRandomTrpFrogPrompt(words)
}
