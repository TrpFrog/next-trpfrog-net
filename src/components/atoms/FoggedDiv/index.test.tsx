import { render, screen } from '@testing-library/react'

import { LOREM_IPSUM } from '@/lib/constants'

import { FoggedDiv } from '.'

describe('FoggedDiv', () => {
  test('snapshot test', () => {
    const view = render(<FoggedDiv height={100} fogHeight={50} />)
    expect(view.asFragment()).toMatchSnapshot()
  })

  test('should render correctly', () => {
    render(<FoggedDiv height={300}>{LOREM_IPSUM}</FoggedDiv>)
    expect(screen.getByText(LOREM_IPSUM)).toBeInTheDocument()
  })
})
