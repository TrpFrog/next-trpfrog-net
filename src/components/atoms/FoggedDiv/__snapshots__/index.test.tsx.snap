// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoggedDiv snapshot test 1`] = `
<DocumentFragment>
  <div
    class="main "
    data-testid="fogged-div"
    style="max-height: 100px;"
  >
    <div
      class="fog"
      data-testid="fogged-div-fog"
    />
  </div>
</DocumentFragment>
`;
